File: src/NumberOfBoomerangs.java
import java.util.HashMap;
import java.util.Map;

public class NumberOfBoomerangs {
    public int numberOfBoomerangs(int[][] points) {
        if (points.length == 0) {
            return 0;
        }

        int boomerangs = 0;
        Map<Integer, Integer> distances = new HashMap<>();

        for (int i = 0; i < points.length; i++) {
            for (int j = 0 ; j < points.length; j++) {
                if (i == j) {
                    continue;
                }

                int distance = getDistance(points[i], points[j]);
                distances.put(distance, distances.getOrDefault(distance, 0) + 1);
            }

            for (int frequency : distances.values()) {
                boomerangs += frequency * (frequency - 1);
            }
            distances.clear();
        }

        return boomerangs;
    }

    private int getDistance(int[] point1, int[] point2) {
        int x_diff = point1[0] - point2[0];
        int y_diff = point1[1] - point2[1];
        return x_diff * x_diff + y_diff * y_diff;
    }
}

File: python/number_of_boomerangs.py
from typing import List


class Solution:
    def get_distance(self, p1: List[int], p2: List[int]) -> int:
        x1, y1 = p1
        x2, y2 = p2
        dx, dy = x2 - x1, y2 - y1
        return dx ** 2 + dy ** 2

    def numberOfBoomerangs(self, points: List[List[int]]) -> int:
        result = 0
        for p1 in points:
            distance_frequency = {}
            for p2 in points:
                distance = self.get_distance(p1, p2)
                distance_frequency[distance] = distance_frequency.get(distance, 0) + 1
            for frequency in distance_frequency.values():
                result += frequency * (frequency - 1)
        return result
